### PR TITLE
ISPN-14813 Memached backpressure support

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-14813

Most of the changes are based on #10934 and #11006.

I ran locally using the `memtier_benchmark`. If anyone is interested in running, I ran with the configuration `memtier_benchmark -p 11222 -4 -P memcache_binary -d 9000 -R --pipeline=10 --test-time=60`, varying the protocol with `memcache_binary` and `memcache_text`, and running one time to warm up and another to collect the info, additionally:

```
4         Threads
50        Connections per thread
60        Seconds
```

Comparison with `--pipeline=1`: https://gist.github.com/jabolina/99ea2e4d8317f36ad12d445bc57672a9
Comparison with `--pipeline=10`: https://gist.github.com/jabolina/6db34b4e1afce874ff7247a9ea9625d3

We can see that `pipeline=1` performs slightly worse but performs better with `pipeline=10`.

I am opening for CI to run, and if anyone wants to take a look already. I am going to check allocations, although I expect a behavior similar to main.

### CODE REVIEW

**Comment by wburns** [comment]
We need to remove all the usages of Unpooled in this new code. RESP and Memcached were built on the idea of a single threaded server. Due to this they have consistency guarantees that multiple commands sent from the same client must be processed in order and with no overlapping. This is done in the RESP server by disabling auto read and not reading any more of the ByteBuf that was passed if a command was not completed synchronously. This allows us some optimizations with reusing a ByteBuf that we cannot do otherwise.

**Comment by jabolina** [comment]
Ok, several changes in this update:

* Not using `Unpooled` in binary;
* Binary allocates a single `BinaryHeader` per channel instead of per operation;
* Text allocates headers per operation if access logging is enabled;
* Text protocol parses the command ID without allocating a string;
* Text protocol allocates a single buffer per channel to parse the arguments instead of slicing the buffer;
* Update the handler for both protocols to reduce the lambda allocation for get and set operations;
* Added replicated tests for binary.

We could update both protocols to avoid allocating the response per operation and do something similar to the `BinaryHeader` with one instance per channel.

**Comment by jabolina** [comment]
Comparison using `memtier_benchmark` with the same setup as in the PR description.

<details>
  <summary>Binary pipeline 1 (-0.0168518%)</summary>
  
### MAIN:
```text
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets        23883.51          ---          ---         0.77878         0.73500         1.31100         6.39900    211007.34
Gets       238818.99       170.35    238648.64         0.75928         0.73500         1.29500         5.24700      9069.54
Waits           0.00          ---          ---             ---             ---             ---             ---          ---
Totals     262702.50       170.35    238648.64         0.76106         0.73500         1.29500         5.37500    220076.88
```

### NEW:
```text
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets        23879.42          ---          ---         0.77591         0.74300         1.27100         4.89500    210971.18
Gets       238778.81       166.68    238612.13         0.75971         0.73500         1.24700         4.03100      9068.01
Waits           0.00          ---          ---             ---             ---             ---             ---          ---
Totals     262658.23       166.68    238612.13         0.76118         0.73500         1.25500         4.07900    220039.19

```
</details>

<details>
  <summary>Text pipeline 1 (-1.32795%)</summary>
  
### MAIN:
```text
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
----------------------------------------------------------------------------------------------------------------------------
Sets        23602.17          ---          ---         0.78859         0.74300         1.31900         5.66300    208360.36
Gets       236002.96       165.46    235837.50         0.76814         0.74300         1.30300         4.60700      7425.86
Waits           0.00          ---          ---             ---             ---             ---             ---          ---
Totals     259605.13       165.46    235837.50         0.77000         0.74300         1.30300         4.70300    215786.22
```

### NEW:
```text
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets        23288.60          ---          ---         0.79703         0.75900         1.31900         5.79100    205592.19
Gets       232869.09       159.15    232709.95         0.77870         0.75100         1.29500         4.54300      7290.98
Waits           0.00          ---          ---             ---             ---             ---             ---          ---
Totals     256157.70       159.15    232709.95         0.78036         0.75100         1.29500         4.63900    212883.16
```
</details>

<details>
  <summary>Binary pipeline 10 (87.9454%)</summary>
  
### MAIN:
```text
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets        65746.46          ---          ---         2.76347         1.59900        19.96700        36.86300    580860.03 
Gets       657447.39      1343.73    656103.67         2.76130         1.59900        19.96700        36.86300     24967.74
Waits           0.00          ---          ---             ---             ---             ---             ---          ---
Totals     723193.85      1343.73    656103.67         2.76150         1.59900        19.96700        36.86300    605827.77
```

### NEW:
```text
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets       123566.07          ---          ---         1.46716         1.22300         7.42300        17.02300   1091687.56
Gets      1235643.75      5215.86   1230427.89         1.46805         1.22300         7.42300        17.02300     46925.78
Waits           0.00          ---          ---             ---             ---             ---             ---          --- 
Totals    1359209.82      5215.86   1230427.89         1.46797         1.22300         7.42300        17.02300   1138613.34
```
</details>

<details>
  <summary>Text pipeline 10 (92.6426%)</summary>
  
### MAIN:
```text
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets        62523.60          ---          ---         2.90514         1.74300        20.73500        37.37500    551959.20
Gets       625219.26      1216.34    624002.92         2.90338         1.74300        20.73500        37.11900     26534.81
Waits           0.00          ---          ---             ---             ---             ---             ---          ---
Totals     687742.86      1216.34    624002.92         2.90354         1.74300        20.73500        37.11900    578494.01
```

### NEW:
```text
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets       120445.63          ---          ---         1.50473         1.28700         6.52700        17.15100   1063295.58
Gets      1204440.05      4832.37   1199607.68         1.50559         1.28700         6.49500        17.15100     73072.51
Waits           0.00          ---          ---             ---             ---             ---             ---          ---
Totals    1324885.68      4832.37   1199607.68         1.50551         1.28700         6.49500        17.15100   1136368.09
```
</details>

The improvement I am comparing is the total ops/sec. The p99 has a smaller value in all executions with the changes.

**Comment by jabolina** [comment]
Still a few failures in integration tests with Memcached, taking a look.

**Comment by jabolina** [comment]
I am running this against mc-crusher and finding some instances the benchmark works with Memcached but not with our impl. So I'll take the chance to fix everything found.

**Comment by jabolina** [comment]
I included another commit with a fix for the `stats` command.

The thing with mc-crusher is that I needed additional configuration. IIUC it establishes a connection and issues one type of command. When we use Memcached text on a single port, the authentication invokes a fake set request, and we identify the protocol with that. With mc-crusher, as it does not have authentication, there are instances we didn't identify the protocol and just closed the connection. To fix that, we need to configure Memcached in a dedicated endpoint.

Because it is a bit more work compared to memtier to get the numbers with mc-crusher, I'll add them here later. But it is ready for review.

File: server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedInboundAdapter.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+   private void flushBufferIfNeeded(ChannelHandlerContext ctx, boolean runOnEventLoop, MemcachedResponse res) {
+      if (outbound != null) {
+         if (runOnEventLoop) {
+            ctx.channel().eventLoop().execute(() -> {
+               if (res != null && progressive != null) {
+                  res.flushed(ctx.writeAndFlush(outbound));
+               } else {
+                  ctx.writeAndFlush(outbound, ctx.voidPromise());
+                  notifyPendingResponses(ctx);
```
Comment (wburns) [inline server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedInboundAdapter.java:59]: Don't pending responses require progressive? It seems inconsistent that progressive is not used in the other if branch but it used here.

File: server/memcached/src/main/java/org/infinispan/server/memcached/binary/BinaryDecoder.java [hunk 2]
```diff
@@ -26,19 +28,21 @@
  **/
 abstract class BinaryDecoder extends MemcachedBaseDecoder {
 
+   protected BinaryHeader singleHeader = new BinaryHeader();
```
Comment (wburns) [inline server/memcached/src/main/java/org/infinispan/server/memcached/binary/BinaryDecoder.java:31]: Unused?

File: server/memcached/src/main/java/org/infinispan/server/memcached/text/TextOpDecoder.java [hunk 3]
```diff
@@ -64,35 +68,48 @@ protected TextOpDecoder(MemcachedServer server, Subject subject) {
...
-      } else {
-         byte[] key = keys.get(0);
-         send(header, cache.getCacheEntryAsync(key).thenApply(entry -> createGetResponse(key, entry, withVersions)));
+
+         CompletionStage<CacheEntry<byte[], byte[]>[]> cs = acs.freeze();
+         if (CompletionStages.isCompletedSuccessfully(cs)) {
+            return send(header, CompletableFuture.completedFuture(createMultiGetResponse(entries)));
+         }
+
+         return send(header, cs.thenApply(unused -> createMultiGetResponse(entries)));
```
Comment (wburns) [inline server/memcached/src/main/java/org/infinispan/server/memcached/text/TextOpDecoder.java:85]:
```suggestion
         return send(header, cs.thenApply(innerEntries -> createMultiGetResponse(innerEntries)));
```

I think this was the original intention of storing the entries in the AggregateCompletionstage.

Also I just noticed that the writing of entries is not synchronized, so I think this is technically an issue that the invoking thread here may not 100% see the value.

File: server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedInboundAdapter.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+         res.writeResponse(obj, allocator);
+         flushBufferIfNeeded(ctx, false, res);
+         resumeAutoRead(ctx);
+      }, ctx.channel().eventLoop());
+   }
+
+   private void registerResponseForLater(MemcachedResponse res) {
+      if (progressive != null) {
+         progressiveEnd++;
+         progressive = progressive.addListener(new MemcachedProgressiveListener(res, progressive));
```
Comment (wburns) [inline server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedInboundAdapter.java:193]: I worry about this being a memory leak... were you able to confirm all of these listeners are not referencing each other? I wonder if instead we should assign progressive to be final and the listeners we register also store the progressive offset in it. Then when we progress it all the listeners that have a value <= can remove themselves and call flush?

**Review by wburns — COMMENTED** [review]
Just a last few things. The progressive is the only thing holding me from just pressing the button though.

File: server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedInboundAdapter.java (hunk 1 above)
Comment (jabolina) [reply server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedInboundAdapter.java]: Updated.

File: server/memcached/src/main/java/org/infinispan/server/memcached/binary/BinaryDecoder.java (hunk 2 above)
Comment (jabolina) [reply server/memcached/src/main/java/org/infinispan/server/memcached/binary/BinaryDecoder.java]: The generated decoder uses it. I added a method so it is more explicit.

File: server/memcached/src/main/java/org/infinispan/server/memcached/text/TextOpDecoder.java (hunk 3 above)
Comment (jabolina) [reply server/memcached/src/main/java/org/infinispan/server/memcached/text/TextOpDecoder.java]: Updated. I am wrapping everything in a synchronized list.

File: server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedInboundAdapter.java (hunk 4 above)
Comment (jabolina) [reply server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedInboundAdapter.java]: I updated with the approach you suggested. Analyzing from jmc and mat does not report a leak. Although, we rely on this only when the access log is enabled, and perf really takes a hit in this case.

**Comment by wburns** [comment]
Integrated into main, thanks @jabolina !

**Pull request merged**